File: 2021/day-3/part-1.php
<?php

// get puzzle input
$file = fopen("data.txt", "r");
$data = explode("\n", fread($file, filesize("data.txt")));
fclose($file);

$gammaRate = "";
$epsilonRate = "";

// For each bit of a line
for ($i = 0; $i < strlen($data[0]); $i++) {
    $sum = 0;

    // Sum the numbers
    foreach ($data as $line) {
        $sum += intval($line[$i]);
    }

    // If the sum is greater than 500 -> append 1 to gammaRate
    if ($sum > sizeof($data) / 2) {
        $gammaRate .= "1";
    } else {
        $gammaRate .= "0";
    }
}

// set epsilon rate by inverting the bits of gammaRate
$epsilonRate = strtr($gammaRate,[1,0]);

echo bindec($gammaRate) * bindec($epsilonRate);

?>

File: 2021/day-3/part-1.py
from data import data

g = ""      # Gamma rate
e = ""      # Epsilon rate

for i in range(len(data[0])):
    sum_of_column = sum([int(x[i]) for x in data])
    if sum_of_column > len(data) / 2:
        g += "1"
    else:
        g += "0"

e = ''.join('1' if x == '0' else '0' for x in g)
print(int(g, 2) * int(e, 2))